safehttp/plugins/fetch_metadata/fetch_metadata.go

```diff
@@ -0,0 +1,155 @@
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fetchmetadata
+
+import (
+	"github.com/google/go-safeweb/safehttp"
+)
+
+// LoggingService is a user-provided service for logging Fetch Metadata policy
+// violations
+type LoggingService interface {
+	Log(ir *safehttp.IncomingRequest)
+}
+
+// Plugin implements Fetch Metadata functionality.
+// See https://www.w3.org/TR/fetch-metadata/ for more details.
+type Plugin struct {
+	// Policy is the Fetch Metadata policy that decides whether a request should
+	// be allowed to pass or be rejected. This will be set to a default Resource
+	// Isolation Policy, but can be changed by the user.
+	//
+	// See https://web.dev/fetch-metadata/ for more information on the default policy
+	Policy func(*safehttp.IncomingRequest, map[string]bool) bool
+	// NavIsolation indicates whether the Navigation Isolation Policy should
+	// also be applied to the request as an additional layer of checking. This
+	// provides a way to mitigate clickjacking and reflected XSS by rejecting
+	// all cross-site navigations unless coming from origins where CORS is
+	// explicitly enabled. This can be used independently from a Resource
+	// Isolation Policy.
+	//
+	// WARNING: This is still an experimental feature and will be disabled by default.
+	NavIsolation bool
+	reportOnly   bool
+	logger       LoggingService
+	allowedCORS  map[string]bool
+}
+
+// NewPlugin creates a new Fetch Metadata plugin in enforce mode using the
+// defaultPolicy and a set of user-provided origins on which CORS is enabled.
+func NewPlugin(origins ...string) *Plugin {
+	m := map[string]bool{}
+	for _, origin := range origins {
+		m[origin] = true
+	}
+	return &Plugin{
+		allowedCORS: m,
+		Policy:      defaultPolicy,
+	}
+}
+
+func defaultPolicy(r *safehttp.IncomingRequest, allowedCORS map[string]bool) bool {
+	h := r.Header
+	switch h.Get("Sec-Fetch-Site") {
+	case "":
+		// Fetch Metadata is not supported by the browser so we allow the
+		// request to pass.
+		return true
+	case "same-origin", "same-site", "none":
+		// The request originated from a site served by your own server
+		// ("same-origin"), a subdomain of your site ("same-site", e.g.
+		// bar.foo.com made a request to foo.com) or was caused by the user
+		// explicitly interacting with the user-agent ("none"). Therefore it is
+		// allowed to pass.
+		return true
+	}
+
+	if m := r.Method(); h.Get("Sec-Fetch-Mode") == "navigate" && (m == safehttp.MethodGet || m == safehttp.MethodHead) {
+		if dest := h.Get("Sec-Fetch-Dest"); dest == "object" || dest == "embed" {
```

Comment:
These two lines are hard to read. I suggest splitting them up and adding comments:

```go
	if method, mode, dest := r.Method(), h.Get("Sec-Fetch-Mode"), h.Get("Sec-Fetch-Dest"); 
		// Allow navigations
		(mode == "navigate" || mode == "nested-navigate") &&
		// On non-state-changing methods
		(method == safehttp.MethodGet || method == safehttp.MethodHead) &&
		// That are not potentially dangerous
		(dest != "object" && dest != "embed") {
			return true
		}
```

That said I think it would even better to have some of this logic factored out of the `if`.

Our internal policy looks more or less like this:
```go
    // (Omitted) read and assign the 3 sec-fetch headers and the method

    if site != "cross-site" {
      // Allow requests from browsers which don't send Fetch Metadata
      // and same-site, same-origin, and browser-initiated requests.
      return true;
    }

    // Allow CORS options requests if neither Mode nor Dest is set.
    // https://github.com/w3c/webappsec-fetch-metadata/issues/35
    // https://bugs.chromium.org/p/chromium/issues/detail?id=979946
    if mode == "" && dest == "" && method == "OPTIONS" {
      return true;
    }

    // Allow navigations from anywhere for non state changing requests.
    // Don't fail-open if mode is not set as e.g. safe browsing doesn't set Fetch Metadata mode.
    if navigationalModes[mode] && navigationalDests[dest] && statePreservingMethods[method] {
      return true;
    }

    return false;
```
Where the three conditions of the last `if` are accesses to maps of type `map[string]bool`

I think this would be a readability boost.

In general if you have a very complicated condition to express I strongly advise to have all nested `return` statements to return one value (`true` in this case) and the last one to return the opposite. It makes it easier to debug.